cudax/include/cuda/experimental/__memory_resource/resource.cuh

```diff
@@ -30,20 +30,38 @@
 namespace cuda::experimental
 {
 
+// Outline to a struct, because resource concept relies on equality comparison, so we can run into a circular dependency
+// when below operator== is taken into consideration for equality comparison trait.
+template <bool _IsDifferent>
+struct __different_non_polymorphic_resources_impl
+{
+  template <class _Resource, class _OtherResource>
+  static constexpr bool __value(_Resource*, _OtherResource*) noexcept
+  {
+    return false;
+  }
+};
+
+template <>
+struct __different_non_polymorphic_resources_impl<true>
+{
+  template <class _Resource, class _OtherResource>
+  static constexpr bool __value(_Resource*, _OtherResource*) noexcept
+  {
+    return _CUDA_VMR::resource<_Resource> && _CUDA_VMR::resource<_OtherResource> && __non_polymorphic<_Resource> && __non_polymorphic<_OtherResource>;
+  }
+};
+
 template <class _Resource, class _OtherResource>
-_CCCL_CONCEPT __comparable_resources = _CCCL_REQUIRES_EXPR((_Resource, _OtherResource))(
-  requires(_CUDA_VMR::__different_resource<_Resource, _OtherResource>),
-  requires(_CUDA_VMR::resource<_Resource>),
-  requires(_CUDA_VMR::resource<_OtherResource>),
-  requires(__non_polymorphic<_Resource>),
-  requires(__non_polymorphic<_OtherResource>));
+_CCCL_CONCEPT __different_non_polymorphic_resources = __different_non_polymorphic_resources_impl<_CUDA_VMR::__different_resource<_Resource, _OtherResource>>::__value(
+  static_cast<_Resource*>(nullptr), static_cast<_OtherResource*>(nullptr));
 
 //! @brief Equality comparison between two resources of different types.
 //! @param __lhs The left-hand side resource.
 //! @param __rhs The right-hand side resource.
 //! @returns Always returns false.
 _CCCL_TEMPLATE(class _Resource, class _OtherResource)
-_CCCL_REQUIRES(__comparable_resources<_Resource, _OtherResource>)
+_CCCL_REQUIRES(__different_non_polymorphic_resources<_Resource, _OtherResource>)
```

Comment:
I believe it should be sufficient to explicitly check that `!is_same_v<_Resource, _OtherResource>`